katas/content/teleportation/Common.qs

```diff
@@ -122,4 +122,61 @@ namespace Kata.Verification {
         Message("Correct.");
         return true;
     }
+
+    operation PrepareAndSendMessage_Reference(qAlice : Qubit, basis : Pauli, state : Bool) : (Bool, Bool) {
+        use qMessage = Qubit();
+        if state {
+            X(qMessage);
+        }
+        if (basis != PauliZ) {
+            H(qMessage);
+        }
+        if (basis == PauliY) {
+            S(qMessage);
+        }
+        let classicalBits = SendMessage_Reference(qAlice, qMessage);
+        Reset(qMessage);
+        return classicalBits;
+    }
+
+    operation ReconstructAndMeasureMessage_Reference(qBob : Qubit, (b1 : Bool, b2 : Bool), basis : Pauli) : Bool {
+        ReconstructMessage_Reference(qBob, (b1, b2));
+        return Measure([basis], [qBob]) == One;
+    }
+
+    // ------------------------------------------------------
+    // Runs teleportation for each state that is to be prepared and
+    // sent by Alice. Success is asserted after each teleportation.
+    // Also repeats for each state several times; this is because
+    // code is expected to take different paths each time since
+    // measurements done by Alice are not deterministic.
+    operation TeleportPreparedStateTestLoop(
+        prepareAndSendMessageOp : ((Qubit, Pauli, Bool) => (Bool, Bool)), 
+        reconstructAndMeasureMessageOp : ((Qubit, (Bool, Bool), Pauli) => Bool)
+        ) : Bool {
+        
+        let messages = [(PauliX, false, "|+>"), 
+                        (PauliX, true, "|->"), 
+                        (PauliY, false, "|i>"), 
+                        (PauliY, true, "|-i>"), 
+                        (PauliZ, false, "|0>"), 
+                        (PauliZ, true, "|1>")];
+        let numRepetitions = 100;
+        use (qAlice, qBob) = (Qubit(), Qubit());
+        for (basis, sentState, stateName) in messages {
+            for j in 1 .. numRepetitions {
+                StatePrep_BellState(qAlice, qBob, 0);
+                let classicalBits = prepareAndSendMessageOp(qAlice, basis, sentState);
+                let receivedState = reconstructAndMeasureMessageOp(qBob, classicalBits, basis);
+                if (sentState != receivedState) {
+                    Message($"Sent and received states were not equal for {stateName}.");
```

Comment:
This begs for an indication of the received state... It's a bit tricky to get the printable form of the incorrect state with these data structures, so let's indicate the basis we're using in this test and hope that with the basis and the basis state it is _not_ the basis state we mean can be deduced
```suggestion
                    Message($"Received incorrect basis state when sending {stateName} in the {basis} basis.");
```